contracts/button/button.rs

```diff
@@ -0,0 +1,399 @@
+}
+
+impl From<InkEnvError> for Error {
+    fn from(e: InkEnvError) -> Self {
+        match e {
+            InkEnvError::Decode(_e) => {
+                Error::ContractCall(String::from("Contract call failed due to Decode error"))
+            }
+            InkEnvError::CalleeTrapped => Error::ContractCall(String::from(
+                "Contract call failed due to CalleeTrapped error",
+            )),
+            InkEnvError::CalleeReverted => Error::ContractCall(String::from(
+                "Contract call failed due to CalleeReverted error",
+            )),
+            InkEnvError::KeyNotFound => Error::ContractCall(String::from(
+                "Contract call failed due to KeyNotFound error",
+            )),
+            InkEnvError::_BelowSubsistenceThreshold => Error::ContractCall(String::from(
+                "Contract call failed due to _BelowSubsistenceThreshold error",
+            )),
+            InkEnvError::TransferFailed => Error::ContractCall(String::from(
+                "Contract call failed due to TransferFailed error",
+            )),
+            InkEnvError::_EndowmentTooLow => Error::ContractCall(String::from(
+                "Contract call failed due to _EndowmentTooLow error",
+            )),
+            InkEnvError::CodeNotFound => Error::ContractCall(String::from(
+                "Contract call failed due to CodeNotFound error",
+            )),
+            InkEnvError::NotCallable => Error::ContractCall(String::from(
+                "Contract call failed due to NotCallable error",
+            )),
+            InkEnvError::Unknown => {
+                Error::ContractCall(String::from("Contract call failed due to Unknown error"))
+            }
+            InkEnvError::LoggingDisabled => Error::ContractCall(String::from(
+                "Contract call failed due to LoggingDisabled error",
+            )),
+            InkEnvError::EcdsaRecoveryFailed => Error::ContractCall(String::from(
+                "Contract call failed due to EcdsaRecoveryFailed error",
+            )),
+            #[cfg(any(feature = "std", test, doc))]
+            InkEnvError::OffChain(_e) => {
+                Error::ContractCall(String::from("Contract call failed due to OffChain error"))
+            }
+        }
+    }
+}
+
+/// Game contracts storage
+#[derive(Debug, SpreadLayout, SpreadAllocate)]
+#[cfg_attr(
+    feature = "std",
+    derive(scale_info::TypeInfo, ink_storage::traits::StorageLayout)
+)]
+pub struct ButtonData {
+    /// How long does TheButton live for?
+    pub button_lifetime: u32,
+    /// Stores a mapping between user accounts and the number of blocks they extended The Buttons life for
+    pub presses: Mapping<AccountId, u32>,
+    /// stores keys to `presses` because Mapping is not an Iterator. Heap-allocated so we might need Map<u32, AccountId> if it grows out of proportion
+    pub press_accounts: Vec<AccountId>,
+    /// stores total sum of user scores
+    pub total_scores: u32,
```

Comment:
I aliased it as a BlockNumber from the `Environment` trait. 
(Scores are denominated in block numbers). The default env defines it as u32, but since we are generic over Environment - it can be easily changed if needed

Reply:
I don't see why do we have to stick with `BlockNumber` semantics here. For custom `EarlyBirdSpecial` with a long deadline, say 1 million blocks (12 days), a couple of thousands presses can easily overflow `u32`

Reply:
it is u64, env is custom for game contracts (the type for BlockNum is the only change)